### PR TITLE
feat: Adds `defn-` and `def-` macros

### PR DESCRIPTION
**Pull request by TimDeve**
Adding these macros as a shortand for declaring a def or defn and making them `hidden` and `private`, useful to keep things internal to a module. Similar to Clojure's [`defn-`](https://clojuredocs.org/clojure.core/defn-).

### CODE REVIEW

**Comment by TimDeve** [comment]
Thinking of adding some example of using `hidden` & `private` in the module section of the [Language Guide](https://github.com/carp-lang/Carp/blob/master/docs/LanguageGuide.md) and introduce these two macros at the same time. Does that sound approriate?

**Comment by scolsen** [comment]
> Thinking of adding some example of using `hidden` & `private` in the module section of the [Language Guide](https://github.com/carp-lang/Carp/blob/master/docs/LanguageGuide.md) and introduce these two macros at the same time. Does that sound approriate?

Sounds great! We should add some tests for these too.

**Comment by TimDeve** [comment]
Yeah I was thinking about the tests yesterday. Would you expect some "error message" tests?

**Comment by hellerve** [comment]
> Yeah I was thinking about the tests yesterday. Would you expect some "error message" tests?

I’d say so! We can use the `test-for-errors` setup.

File: core/Macros.carp [hunk 1]
```diff
@@ -133,12 +133,19 @@
   (let [v (if (= (length explanation) 0) true (car explanation))]
     (eval (list 'meta-set! name "deprecated" v))))
 
-(defmodule Dynamic
-
-  (defndynamic quoted [x]
```
Comment (hellerve) [inline core/Macros.carp:138]: Where did this go?

**Comment by TimDeve** [comment]
> I’d say so! We can use the test-for-errors setup.

Added the tests a few minutes before your message. Doing docs next.

**Comment by TimDeve** [comment]
Took me weeks to find the time to make this simple PR but I think it's finally ready for review 😄

File: docs/LanguageGuide.md [hunk 2]
```diff
@@ -227,6 +156,88 @@ These can only be used at the REPL and during macro evaluation. Here's a subset
 
 To see all functions available in the `Dynamic` module, enter `(info Dynamic)` at the REPL.
 
+
+### Structs
+Any structure type defined in Carp has an init method that can be used to create a new instance. It must be called with all the arguments in the order they are defined.
```
Comment (hellerve) [inline docs/LanguageGuide.md:161]: Could we wrap `init` in backticks here?

Also, maybe "in the order in which they are defined" or "in definition order" might be better?

File: docs/LanguageGuide.md [hunk 3]
```diff
@@ -227,6 +156,88 @@ These can only be used at the REPL and during macro evaluation. Here's a subset
...
+```
+
+**Data:**
+```clojure
+(deftype (Either a b)
+  (Left [a])
+  (Right [b]))
+```
+
+A Variant can be created with the same syntax as call expression:
```
Comment (hellerve) [inline docs/LanguageGuide.md:200]: `A variant can be created through a function call`?

File: docs/LanguageGuide.md [hunk 4]
```diff
@@ -227,6 +156,88 @@ These can only be used at the REPL and during macro evaluation. Here's a subset
...
+    (Either.Right b) b))
+
+(with MyEnum
+  ;; You can give a generic "otherwise" statement as well
+  (match myenum
+    (Kind1) (logic1)
+    _ (logic-other)))
+```
+
+Note that match works with *values* (not references) takes ownership over the value being matched on. If you instead want to match on a reference, you can use `match-ref`:
```
Comment (hellerve) [inline docs/LanguageGuide.md:231]: I think backticks around `match` would be good here, and that there is an `and` missing after the parentheses.

**Review by hellerve — COMMENTED** [review]
I realized about halfway through that the docs I was annotating weren’t new but rearranged. My comments still hold, I think 😆

**Comment by hellerve** [comment]
I think it would be good to use these forms in stdlib in a follow-up PR, since this could remove a lot of code and test the forms at the same time.

**Comment by TimDeve** [comment]
The follow up PR sounds like a good idea. About the comments you've added I was thinking of making a refractor PR for the language guide, the style is not very consistent and some of it seems straight up wrong with the changes in the language since it was written, so I could put your suggested changes it that.

**Comment by hellerve** [comment]
> The follow up PR sounds like a good idea. About the comments you've added I was thinking of making a refractor PR for the language guide, the style is not very consistent and some of it seems straight up wrong with the changes in the language since it was written, so I could put your suggested changes it that.

Sounds like a good plan! I’ll be happy to help and/or review :)

**Comment by TimDeve** [comment]
Once/if this is merged I'll open a draft PR with your suggested changes and we can comment on it on what we add to that.

**Comment by eriksvedang** [comment]
Looks good! The changes to the language guide is a bit hard to follow though, has anything been deleted or have things just moved around?

**Comment by TimDeve** [comment]
Sorry the merged diff looks really messy, it's easier to follow if you go commit by commit. I moved the module and interface section after the struct section first, and then I added a section about `private` and `hidden` in the module section.

**Comment by eriksvedang** [comment]
@TimDeve Alright, thanks I'll do that!

**Review by eriksvedang — APPROVED** [review]
@hellerve had some comments that you can choose to address, apart from that I think we're good to go!

**Comment by eriksvedang** [comment]
@TimDeve Maybe you already discussed that and this should just go in as-is now?

**Comment by TimDeve** [comment]
Yeah the comment where on existing docs so we're going to address them in a different PR.

**Comment by eriksvedang** [comment]
Great work!
